### PR TITLE
[OpenGL] Support 'ti.asm' to insert embed GLSL codes (experimental)

### PR DESCRIPTION
**Pull request by archibate**
Related issue = #1378

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----
This will enable users to utilize backend features, e.g.: [`ti.asm('memoryBarrier()')`](http://docs.gl/sl4/memoryBarrier).
@k-ye Feel free to add a Metal one after this is merged :)

### CODE REVIEW

File: misc/demo_embed_asm.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import taichi as ti
```
Comment (yuanming-hu) [inline misc/demo_embed_asm.py:1]: Could you rename this file to `misc/demo_opengl_embed_asm.py` or simply make it a test?

File: taichi/backends/opengl/codegen_opengl.cpp [hunk 2]
```diff
@@ -644,6 +644,36 @@ class KernelGen : public IRVisitor {
...
+          num *= 10;
+          num += format[i] - '0';
+        }
+        auto args = (c == '%') ? stmt->arg_stmts : stmt->output_stmts;
+        TI_ASSERT_INFO(num < args.size(), "{}{} out of argument range {}", c,
+                       num, args.size());
+        source += args[num]->short_name();
+        continue;
+      }
+      source.push_back(c);
```
Comment (yuanming-hu) [inline taichi/backends/opengl/codegen_opengl.cpp:671]:
```suggestion
      } else {
        source.push_back(c);
      }
```

File: taichi/backends/opengl/codegen_opengl.cpp [hunk 3]
```diff
@@ -644,6 +644,36 @@ class KernelGen : public IRVisitor {
...
+          if (!::isdigit(format[i])) {
+            i -= 1;
+            break;
+          }
+          num *= 10;
+          num += format[i] - '0';
+        }
+        auto args = (c == '%') ? stmt->arg_stmts : stmt->output_stmts;
+        TI_ASSERT_INFO(num < args.size(), "{}{} out of argument range {}", c,
+                       num, args.size());
```
Comment (yuanming-hu) [inline taichi/backends/opengl/codegen_opengl.cpp:667]: Better specify if the arg is an input argument or output variable?

File: misc/demo_embed_asm.py (hunk 1 above)
Comment (archibate) [reply misc/demo_embed_asm.py]: Done, using `ti.extension.extfunc`, we may also write a extfunc test on CPU later.

**Pull request merged**